python/pyspark/sql/tests/test_streaming.py

```diff
@@ -592,6 +592,18 @@ def collectBatch(df, id):
             if q:
                 q.stop()
 
+    def test_streaming_foreachBatch_graceful_stop(self):
+        # SPARK-39218: Make foreachBatch streaming query stop gracefully
+        def func(batch_df, _):
+            time.sleep(10)
+            batch_df.count()
+
+        q = self.spark.readStream.format("rate").load().writeStream.foreachBatch(func).start()
+        time.sleep(5)
+        q.stop()
+        time.sleep(15)  # Wait enough for the exception to be propagated if exists.
+        self.assertIsNone(q.exception(), "No exception has to be propagated.")
```

Comment:
I just took a simpler path. I can use the test source approach like `ThrowingExceptionInCreateSource` but that would require more complicated test cases like https://github.com/apache/spark/blob/master/python/pyspark/sql/tests/test_dataframe.py#L1245-L1266.

The test cases here allow false positive but won't be flaky at least.